### ISSUE TITLE
Auto-deriving FromHttpApiData and ToHttpApiData

### ISSUE THREAD

**Issue by tysonzero** [haskell-servant/servant#1014]
Is there any chance it will be made possible to auto-derive `FromHttpApiData` and `ToHttpApiData`, similar to how we can auto derive `FromJSON` and `ToJSON` in Aeson. I would enjoy being able to easily use more interesting types in my URLs.

**Comment by phadej** [comment]
this belongs to `http-api-data`.

- How would it work for anything else than textual data?
- How it would even guess a format of textual data, if parsing can fail?
- Doesn't `GeneralizedNewtypeDeriving` work for `newtype`s?
- When we'll have `DerivingVia`, something like https://github.com/futurice/haskell-futurice-prelude/blob/d112e0cffc8beaf2082762e5ae53f0071514ab97/src/Futurice/Generics/Textual.hs would make sense to be packaged, not sure if it's worth right now.

**Comment by phadej** [comment]
That said, if you have few types in mind, which `FromHttpApiData` ought to be derived, please give examples!

**Comment by tysonzero** [comment]
> How would it work for anything else than textual data?

I was assuming that an approach similar to Aeson's would work, but I could be missing some important aspects of URIs that make such a thing not possible.

> How it would even guess a format of textual data, if parsing can fail?

When you make a request it would presumably try the endpoints one by one until one succeeds, I was under the impression that it was already done in a way similar to that.

> Doesn't GeneralizedNewtypeDeriving work for newtypes?

GeneralizedNewtypeDeriving doesn't help me for my case, as my type is not a newtype, it is a custom record. It also seems like tuples do not have a `(From|To)HttpApiData` instance so I can't use a newtype over those.

**Comment by alpmestan** [comment]
Well, FromHttpApiData/ToHttpApiData let you give textual encodings of data types so as to put the result in path fragments or in query parameter values. There isn't a standard scheme for doing this for "about any data you have around", unlike JSON for which we have a natural mapping between its records/arrays/prmitive types and Haskell's. So it is indeed hard to provide a generic implementation of those things for all Haskell data types (that derive e.g `Generic`), because what should

``` haskell
data Foo = Foo Int Char String Double Double (Maybe FilePath)
```

look like when given as the value of a query parameter?

**Comment by tysonzero** [comment]
I would guess it might look something like a URL escaped `Foo 5 'a' "foo" 6.5 2.3 (Just "foo")`.

I would be fine with a URL escaped Show/Read for everything, although I can appreciate why the quotes might not be desired for strings themselves.

I also realize that with such an approach relying on non-collision of URIs is less safe, as nominally different types might overlap in their string representation (safe with Read/Show since generally you know exactly what type you are looking for with no "fall back"), so it should probably be discouraged / statically prevented to have URIs that differ solely in the `FromHttpApiData` instance.

**Comment by alpmestan** [comment]
Right, this is not really that common a scenario for reasons including the ones you mention. Does this discussion answer your initial question?

**Comment by tysonzero** [comment]
I honestly do still like the idea of being able to encode arbitrary types into the URL via deriving. My specific use case is that I have a lot of list endpoints that have relatively complex filtering and ordering options, but I would like to keep using a Get and not have to switch to Post + RequestBody.

It seems like having the only difference between two routes be the type of a capture, or a capture vs a string (at least where the string doesn't come first anyway), is kind of a sketchy antipattern anyway. So it still seems like the upsides of easy complex types in the URL (which are far from uncommon in general) outweigh the downsides.

I think just having the generic/derived implementation just require and use `Show` / `Read` would probably work ok.

**Comment by phadej** [comment]
TL;DR you don't need that functionality to exist in `servant` or/and `http-api-data`. Please present an evidence that it's a good idea in general case, and how the URLs would look like etc. Than we might add **your** implementation to proper place.

---

URLs lke

```
http://example.com/Foo%2042%20True%20%22String%22
```

Anyway, you can do:

```haskell
genericToUrlPieceViaShow :: Show a =>  a -> Text
genericToUrlPieceViaShow = Text.pack . show

data MyData = ...

instance ToHttpApiData MyData where
    toUrlPiece = genericToUrlPieceViaShow
```

but having that as **default** implementation for `ToHttpApiData` class (or even as a helper in the library) is IMO quite clear *wontfix*.

---

The more servant approach to *many HTTP terms, but really single "Thing"*, could go into something like Generics based args:

```haskell
type API = "endpoint" :> GCombinators Params :> Get '[JSON] Response

data Params param = Params
    { paramNeedle    :: param :- QueryParam "needle" Text
    , paramDirection :: param :- QueryParam "direction" Dir
    }
```

and I would like to see someone try to implement that.

**Comment by tysonzero** [comment]
I guess my "evidence" for something like that to exist in the library is simply that it allows you to do:

    data Foo = Foo { bar :: Bar, baz :: Baz } deriving (Generic, FromHttpApiData, ToHttpApiData)

or:

    data Foo = Foo { bar :: Bar, baz :: Baz } deriving (Show, Read, FromHttpApiData, ToHttpApiData)

And have everything "just work", with as far as I can tell no real downsides, since url parts are properly escaped and all that. Rather than having to manually specify the instances and manually verify that they are a true bijection. It's similar to why Aeson lets you auto0derive instances, and why Read and Show can be automatically derived.

The URLs would probably just look like their Show/Read instances but URL escaped, unless anyone else has a different approach that they would prefer. All I care about is being able to quickly derive a bijection and not have to think about it, the URIs can be crazy ugly and it won't really bother me.

I thought about the multiple terms -> one thing approach and my main issue with using it much is that it is totally possible for parameters to end up colliding and causing strange and annoying issues, which cannot happen with the single term approach (unless you try and have two routes that only differ based on their `Capture` type, which IMO is an antipattern).

**Comment by phadej** [comment]
@tysonzero Well, I gave you a code to write `ToHttpApiData` instance using `Show`, `Read` is similar (use `readMaybe`). With *DerivingVia* (GHC-8.6) you can make it more convinient

```haskell
data Foo = Foo ...
  deriving stock (Show)
  deriving (ToHttpApiData) via (ToHttpApiDataShow Foo)
```

Having that as default generic is **wontfix*.

**Comment by phadej** [comment]
And to be even more clearer: in many cases aeson generically derives instances which one would write manually themselves. Show/Read based To/FromHttpApiData aren't that; and are "usable" (IMO) only if both client and server are Haskell, at which point there are multitude of options.
